### ISSUE TITLE
Help on simple convolution

### ISSUE THREAD

**Issue by Rares14324** [ARM-software/ComputeLibrary#620]
**Output of 'strings libarm_compute.so | grep arm_compute_version':**
arm_compute_version=v18.11 Build options: {'embed_kernels': '1', 'onpencl': '1', 'arch': 'x86_64', 'neon': '0', 'build': 'native', 'debug': '1', 'os': 'linux', 'Werror': '0'} Git hash=3d2d44ef55ab6b08afda8be48301ce3c55c7bc67
**Platform:**
x64
**Operating System:**
Ubuntu 16.04 LTS

**Problem description:**
Good day,
I'm trying to implement a custom convolution network using this library because it needs to run on an Android device. I have managed to compile the library for linux using openCL in order to have a desktop work environment before final testing on the device. The network is "ported" from tensorflow however I find myself in confusion when using this library to build up the tensorflow network and would appreciate some advice/help on the first layer layer (which I'll explain) of the network to clarify my understanding.

At this [link](https://imgur.com/a/rSPh4Zr)  you can see highlighted the convolution operation along with the convolution kernel from the dump scripts (which was slightly altered to print the kernel to screen). The network (simplified only for this first layer) is attached (DemoNetwork.h) along with relevant data
In my main I use following code to run a test and check that everything is in order:
```
#include <arm_compute/runtime/CL/functions/CLActivationLayer.h>
#include <arm_compute/runtime/CL/functions/CLConvolutionLayer.h>
#include <arm_compute/runtime/CL/CLSubTensor.h>
#include <tests/CL/CLAccessor.h>
#include <arm_compute/core/Types.h>
#include <arm_compute/runtime/CL/CLFunctions.h>
#include <arm_compute/runtime/CL/CLScheduler.h>
using CLDemoModel = DemoNetwork<CLTensor,
      arm_compute::test::CLAccessor,
      CLActivationLayer,
      CLConvolutionLayer,
      CLSubTensor>;
int main(int argc, char* argv[])
{
    CLDemoModel testInitialSS;
    arm_compute::CLScheduler::get().default_init();
    testInitialSS.init();
    testInitialSS.buildNetwork();
    testInitialSS.allocate();
    testInitialSS.fill();
    std::array<float, 135*257> testInput;
    std::vector<float> testOut;
    testInitialSS.feed(testInput);
    testInitialSS.run();
    testInitialSS.sync();
    testInitialSS.getPrediction(testOut);
    testInitialSS.clear();
}
```
Using this configuration I get all zeros inside testOut which means my understanding about the parity of tensorflow functions and compute-arm function is lacking at the data layout. Changing the dimensions leads to other checks failing inside the library.
My question would be given this concrete example how does compute-arm expect the data to be presented for the inputs, for the kernels and bias in order to have equivalence to the tensorflow one ?
(the weights are given by tensorflow as shape (3,3,1,128) but using that in arm_compute::TensorShape(3U, 3U, 1U, 128U) for m_convolutionKernelForInput gives 
in validate src/runtime/CL/functions/CLGEMMConvolutionLayer.cpp:379: (weights->dimension(idx_channel) * num_groups) != input->dimension(idx_channel) ). 

I apologize if it ended up being a huge post and would appreciate any help/guidance you can provide.
Some other notes: I can't use graph API because I have residuals along in my network (and need SubTensor support for implementing them). Inside the sources.zip you can find input data and a screenshot with the parameters from tf.layers.conv2d call from tensorflow. <br/>
Thank-you.

[sources.zip](https://github.com/ARM-software/ComputeLibrary/files/2772735/sources.zip)

**Comment by Rares14324** [comment]
I managed to get pas the initial problem. The zeros were caused by forgetting to pass the alfa parameter to the activation function which was linear (I feel ashamed for missing it). What still gives me bad results is the fact that I don't quite understand the relation between arm compute dimension of tensors and what I get when dumping the parameters. To give an example:
```
        m_convolutionKernel[0].allocator()->init(arm_compute::TensorInfo(
                                                     arm_compute::TensorShape(128U, 3U,3U,128U),
                                                     1,
                                                     arm_compute::DataType::F32));
        m_convolutionKernel[0].info()->set_data_layout(arm_compute::DataLayout::NHWC);
```
These are the weights for an 2d convolution. I interpret from a vector logical point as int weights[128][3][3][128] (disregarding implementation padding) Unfortunately when saving the relevant tensorflow weights it gives me a shape of (3,3,128,128). I don't want to modify arm compute's shape of the tensor because it fails the checks inside the  network. To accommodate it I make a transpose inside the script for this variable like so:
```
# toWrite.shape=(3,3,128,128)
toWrite = np.transpose(toWrite, (2,0,1,3))
# write to file toWrite.flaltten(), how it is written doesn't matter, loading inside c++ is correct as well

```
Assuming that I choose the right axes and that both have the same data layout NHWC, is the following code to load the weight correct ? (disregarding efficiency):
```
bool loadWeights(TensorType& tensor, unsigned char* rawData, unsigned int rawDataSize)
        {
            Accessor accessTensor(tensor);
            int numDims = accessTensor.shape().num_dimensions();
            if (numDims == 4)
            {
                arm_compute::Window inputWindow;
                inputWindow.use_tensor_dimensions(accessTensor.shape());
                arm_compute::Iterator tensorIt(&tensor, inputWindow);
                const std::size_t dim1 = accessTensor.shape()[0];
                const std::size_t dim2 = accessTensor.shape()[1];
                const std::size_t dim3 = accessTensor.shape()[2];
                const std::size_t dim4 = accessTensor.shape()[3];
                // assume a float array
                const float* rawDataAsFloat = reinterpret_cast<float*>(&rawData[0]);
                arm_compute::execute_window_loop(inputWindow, [&](const arm_compute::Coordinates & id)
                {
                    int idx = id[0] * dim2 * dim3 * dim4 + id[2] * dim3 * dim4 + id[1] * dim4 + id[3];
                    *(reinterpret_cast<float*>(tensorIt.ptr())) = rawDataAsFloat[idx];
                }, tensorIt);
```
This is what I cannot fully understand because the window iterates in a reverse order than my understanding of  the tensor shape.
Any help is greatly appreciated.

**Comment by Rares14324** [comment]
I'm updating this issue since it was a misunderstanding on my part. My current understanding is when defining a tensor like this:
`testTensor.allocator()->init(arm_compute::TensorInfo(
                                                     arm_compute::TensorShape(127U, 2U,3U,128U),
                                                     1,
                                                     arm_compute::DataType::F32));
`
the equivalent numpy declaration would be float testTensor[128][3][2][127] from a memory standpoint.
When I was dealing with convolutions arm-compute expected different transposed weight tensors than the dump script provided. In my particular case I was confused because the shape was (3,3,128,128) and it was hard to tell which axis was transposed for arm-compute to not signal errors. This coupled with the fact that I load the weights from header files as opposed to .npy files caused incorrect weights to be loaded. Since I hadn't caught on to this in the first place I suspected the problem in some other place leading to this question.
Sorry if I spammed the issue section and hope that others may learn from this. After correctly adjusting the indexing in the load weights code to account for the correct axis which was transposed I was able to get correct results up to the 5th digits with the reference network.
